### PR TITLE
Add doc about WebAssembly module remote load.

### PR DESCRIPTION
**Pull request by bianpengyuan**
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by ericvn** [comment]
Can we get approval from the @istio/wg-policies-and-telemetry-maintainers.

**Review by douglas-reid — APPROVED** [review]
Content approval. I'll let the docs maintainers cover edit passes, etc.

File: content/en/docs/ops/configuration/extensibility/_index.md
Comment (ericvn) [inline content/en/docs/ops/configuration/extensibility/_index.md:4]: This weight is the same as Observability. It should changed so that the ordering is correct. If you want it just before or after `Observability` you could make it 35 or 45.

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:24]: Last sentence here doesn't make sense, I can't quite figure out what it is trying to say. What is the fetch timeout preventing?

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:52]:
```suggestion
The second `EnvoyFilter` provides configuration for the filter, which is an `EXTENSION_CONFIG` patch and will be distributed to the proxy as an Envoy [`Extension Configuration`](https://www.envoyproxy.io/docs/envoy/latest/configuration/overview/extension) (ECDS) resource.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:53]:
```suggestion
Once this update reaches the Istio agent, the agent will download the Wasm module and store it in the local file system.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:54]:
```suggestion
If the download fails, the agent will reject the `Extension Configuration` update to prevent invalid Wasm filter configuration from reaching the Envoy proxy.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:55]:
```suggestion
The important parts of this configuration are:
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:58]:
```suggestion
- Wasm extension configuration, which is a JSON string that is consumed by the Wasm extension.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:108]: I don't understand what this is trying to tell the user, please rephrase.

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:112]:
```suggestion
The Istio agent will only intercept and download remote Wasm modules configured via ECDS resources.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:113]:
```suggestion
To disable ECDS interception and Wasm downloading in the Istio agent, set the `ISTIO_AGENT_ENABLE_WASM_REMOTE_LOAD_CONVERSION` environment variable to `false`.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:123]:
```suggestion
There are several known limitations with this module distribution mechanism, which will be addressed in future releases:
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:125]:
```suggestion
- Envoy's extension configuration discovery service only supports HTTP filters.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:126]:
```suggestion
- Modules can only be downloaded through HTTP/HTTPS.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:128]:
```suggestion
## Monitoring Wasm Module Distribution Failures
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:130]:
```suggestion
There are several stats which track the distribution status of remote Wasm modules.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:134]:
```suggestion
- `istio_agent_wasm_cache_lookup_count`: number of Wasm remote fetch cache lookups.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:135]:
```suggestion
- `istio_agent_wasm_cache_entries`: number of Wasm config conversions and results, including success, no remote load, marshal failure, remote fetch failure, and miss remote fetch hint.
```

File: content/en/docs/ops/configuration/extensibility/_index.md
Comment (bianpengyuan) [reply content/en/docs/ops/configuration/extensibility/_index.md]: Updated. Thanks!

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:136]:
```suggestion
- `istio_agent_wasm_config_conversion_duration_bucket`: Total time in milliseconds istio-agent spends on config conversion for Wasm modules.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:139]:
```suggestion
If a Wasm filter configuration is rejected, either due to download failure or other reasons, istiod will also emit `pilot_total_xds_rejects` with the type label `type.googleapis.com/envoy.config.core.v3.TypedExtensionConfig`.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:143]:
```suggestion
To learn more about Wasm module development, please refer to the guides provided in the [`istio-ecosystem/wasm-extensions` repository](https://github.com/istio-ecosystem/wasm-extensions),
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:144]:
```suggestion
which is maintained by the Istio community and used to develop Istio's Telemetry Wasm extension:
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:147]:
```suggestion
- [Write unit tests for C++ Wasm extensions](https://github.com/istio-ecosystem/wasm-extensions/blob/master/doc/write-cpp-unit-test.md)
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:148]:
```suggestion
- [Write integration tests for Wasm extensions](https://github.com/istio-ecosystem/wasm-extensions/blob/master/doc/write-integration-test.md)
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (bianpengyuan) [reply content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md]: I broke the explanation into two parts. First part explains what this option does, and the other part explains why this option is needed below with the second `EnvoyFilter` resource. Hope this could make it more clear.

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (bianpengyuan) [reply content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md]: Removed. I didn't know there is a way to version the yaml file in the doc so wanted to emphasize that this is only tested with Istio 1.9 release. But later I find out there is actually a way to version the EnvoyFilter with `{{< istio_version >}}` substitution, so I don't think we need this note any more.

**Comment by bianpengyuan** [comment]
@smawson @ericvn thanks a lot for the review! Please take another look.

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:3]:
```suggestion
description: Describes how to make remote WebAssembly modules available in the mesh.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:54]:
```suggestion
If the download fails, the agent will reject the `Extension Configuration` update to prevent invalid Wasm filter configuration from reaching the Envoy proxy.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (smawson) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:55]:
```suggestion
Because of this protection, with the initial fetch timeout being set to 0, the listener update will not become effective and an invalid Wasm filter will not disturb traffic.
```

**Review by smawson — COMMENTED** [review]
Fixes look good, a couple more minor edits.

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (dgn) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:87]: should we maybe add this parametrized URL as a comment and use the `.wasm` binary for 1.9 here? we can scope the EnvoyFilter to proxies with version 1.8. That would allow us to use the sha256 as well and establish it as a best practice

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (dgn) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:24]:
```suggestion
With the first `EnvoyFilter`, an HTTP filter will be injected into gateway proxies. It is configured to request the extension configuration named `istio.basic_auth` from `ads` (i.e. Aggregated Discovery Service), which is the same configuration source that Istiod uses to provide all other configuration resources. Within the configuration source, the initial fetch timeout is set to `0s`, which means that when the Envoy proxy processes a listener update with this filter, it will wait indefinitely for the first extension configuration update before accepting requests with this listener.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (dgn) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:18]: should we mention that this only works on `master` and 1.9?

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (dgn) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:108]:
```suggestion
This feature is enabled by default in Istio 1.9. To disable ECDS interception and Wasm downloading in the Istio agent, set the `ISTIO_AGENT_ENABLE_WASM_REMOTE_LOAD_CONVERSION` environment variable to `false`.
```

**Review by dgn — APPROVED** [review]
lgtm, just a few comments. I assume this is planned for publishing after 1.9 goes live?

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (ericvn) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md]:
```suggestion
description: Wow to make remote WebAssembly modules available in the mesh for extensibility.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (frankbu) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:52]:
```suggestion
The second `EnvoyFilter` provides configuration for the filter, which is an `EXTENSION_CONFIG` patch and will be distributed to the proxy as an Envoy [Extension Configuration Discovery Service](https://www.envoyproxy.io/docs/envoy/latest/configuration/overview/extension) (ECDS) resource.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (frankbu) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:54]:
```suggestion
If the download fails, the agent will reject the ECDS update to prevent invalid Wasm filter configuration from reaching the Envoy proxy.
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (frankbu) [inline content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:10]: If this is not an experimental feature, it is applicable., so there should be a test added asap.

```suggestion
test: no
```

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (frankbu) [reply content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:18]: I suppose there could be some mention that this feature is new, not available in releases prior to 1.9. For example, on line 16: `Starting in version 1.9, Istio makes this possible by allowing ...`

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (frankbu) [reply content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:108]: These are the 1.9 docs, so I would leave off the `in Istio 1.9`. It could just say `This feature is enabled by default. To dsable ECDS ...`

File: content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md
Comment (bianpengyuan) [reply content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md:87]: I am trying to avoid remembering updating this doc for every release, since it is really easy to forget. I've update the comment to make it clear that checksum is highly recommended.

**Comment by bianpengyuan** [comment]
@dgn thanks for reviewing. yes, this will publish with 1.9.

**Comment by bianpengyuan** [comment]
@ericvn @frankbu @smawson Can I get another review on this change? thanks!

**Review by ericvn — APPROVED** [review]

**Pull request merged**
